### PR TITLE
[INTERNAL] rework documentation of union_composition

### LINKED ISSUE #60: union_composition clean-up
depends on #58 
  * [x] #68, streamline documentation, in particular: 
    * [x] add `\param` documentation to all functions
    * [x] don't use `namespace seqan3::detail::union_alphabet` since it doesn't appear in doxygen
    * [x] don't use `\privatesection` and `\publicsection` outside of class scope (it doesn't have any effect there)
    * [x] add `\ingroup composition` where needed!
  * [x] #69, streamline tests, in particular
    * [x] remove redundancy, add to generic test cases
    * [x] don't test `_value`
    * [x] tests may only depend on previously tested behaviour, there must be no circular dependencies
    * [x] don't use static_asserts, instead use TESTs
  * [x] #70, add `noexcept` where adequate

if you feel that these issues are not related, feel free to make separate PRs (i.e. not for everything, but for test VS doc for example) :smile:

### LINKED ISSUE #68: union_composition clean-up: documentation
This is a detail/discussion ticket of the meta ticket #60 

 * [x] add `\param` documentation to all functions
 * [x] don't use `namespace seqan3::detail::union_alphabet` since it doesn't appear in doxygen
 * [x] don't use `\privatesection` and `\publicsection` outside of class scope (it doesn't have any effect there)
 * [x] add `\ingroup composition` where needed!

### PR DESCRIPTION
**Pull request by marehr**
* don't use `namespace seqan3::detail::union_alphabet` since it doesn't
appear in doxygen
* don't use `\privatesection` and `\publicsection` outside of class
scope (it doesn't have any effect there)
* add `\ingroup composition` where needed!
* fix testcases of union_composition
* add \param documentation to union_composition

resolves #68 and resolves #60

### CODE REVIEW

File: include/seqan3/alphabet/composition/union_composition.hpp [hunk 1]
```diff
@@ -277,27 +112,55 @@ namespace seqan3
...
+     * ```cpp
+     * using union_t = union_composition<dna4, gap>;
+     *
+     * static_assert(union_t::has_type<dna4>(), "should be true");
+     * static_assert(union_t::has_type<gap>(), "should be true");
+     * static_assert(!union_t::has_type<dna5>(), "should be false");
+     * ```
      */
     template <typename alphabet_t>
     static constexpr bool has_type() noexcept
```
Comment (h-2) [inline include/seqan3/alphabet/composition/union_composition.hpp]: maybe rename `has_type()` to `holds_alternative` to be in line with std::variant's interface?

File: include/seqan3/alphabet/composition/union_composition.hpp [hunk 2]
```diff
@@ -340,6 +204,7 @@ class union_composition
     /*!\brief Construction via a value of reoccurring alphabets
```
Comment (h-2) [inline include/seqan3/alphabet/composition/union_composition.hpp]: brief statement (and all other ones actually as well) should end in a `.`

File: include/seqan3/alphabet/composition/union_composition.hpp [hunk 3]
```diff
@@ -277,27 +112,55 @@ namespace seqan3
...
+     * assert(sum == 10);
+     * ```
+     */
+    template <typename ...alphabets_t>
+    //!\cond
+        requires (alphabet_concept<alphabets_t> && ...)
+    //!\endcond
+    static constexpr auto alphabet_sum_size_v =
+        (static_cast<size_t>(alphabet_size_v<alphabets_t>) + ... + static_cast<size_t>(0));
+
```
Comment (h-2) [inline include/seqan3/alphabet/composition/union_composition.hpp]: can't we move this down into the other protected section?

File: include/seqan3/alphabet/composition/union_composition.hpp [hunk 4]
```diff
@@ -51,177 +51,12 @@
...
-
 namespace seqan3
 {
 
 /*!\brief A composition that merges different regular alphabets as a single alphabet.
  * \ingroup composition
- * \tparam first_alphabet_type Type of the first letter, e.g. dna4; must satisfy seqan3::alphabet_concept.
- * \tparam alphabet_types Types of further letters; must satisfy seqan3::alphabet_concept.
+ * \tparam ...alphabet_types Types of further letters; must satisfy seqan3::alphabet_concept, e.g. dna4.
  *
```
Comment (h-2) [inline include/seqan3/alphabet/composition/union_composition.hpp]: please add 
`\implements seqan3::alphabet_concept`

File: include/seqan3/alphabet/composition/union_composition.hpp [hunk 5]
```diff
@@ -461,39 +329,200 @@ class union_composition
...
-    static constexpr auto value_to_char =
-        detail::union_composition::value_to_char_table<char_type, first_alphabet_type, alphabet_types...>();
+    static constexpr auto value_to_char = value_to_char_table();
 
     //!\brief Compile-time generated lookup table which maps the char to rank
-    //!\hideinitializer
     //!\sa char_to_value_table
-    static constexpr auto char_to_value =
-        detail::union_composition::char_to_value_table<char_type, first_alphabet_type, alphabet_types...>();
+    static constexpr auto char_to_value = char_to_value_table();
```
Comment (h-2) [inline include/seqan3/alphabet/composition/union_composition.hpp]: why not use the above functions as in-place initialisers in form of a lambda?

Comment (h-2) [inline include/seqan3/alphabet/composition/union_composition.hpp]: It would reduce the number of entities that you need to keep track of and and document...

File: include/seqan3/alphabet/composition/union_composition.hpp (hunk 1 above)
Comment (marehr) [inline include/seqan3/alphabet/composition/union_composition.hpp]: It's not exactly the same, `holds_alternative` is true, if the given type and the last assigned type matches.

`has_type` is true, if the given type is within the template pack of the class.

But, I'm also not satisfied with the name.. Maybe `is_base_type` is better.

File: include/seqan3/alphabet/composition/union_composition.hpp [hunk 6]
```diff
@@ -277,27 +112,55 @@ namespace seqan3
...
     template <typename alphabet_t>
     static constexpr bool has_type() noexcept
     {
-        return meta::in<meta::list<first_alphabet_type, alphabet_types...>, alphabet_t>::value;
+        return meta::in<meta::list<alphabet_types...>, alphabet_t>::value;
     }
 
     //!\brief The size of the alphabet, i.e. the number of different values it can take.
-    static constexpr size_t value_size = (alphabet_types::value_size + ... + first_alphabet_type::value_size);
+    static constexpr size_t value_size = alphabet_sum_size_v<alphabet_types...>;
```
Comment (marehr) [inline include/seqan3/alphabet/composition/union_composition.hpp]: > can't we move this down into the other protected section?

It is used here, so I would need to forward that anyway, so I just define it.

File: include/seqan3/alphabet/composition/union_composition.hpp (hunk 5 above)
Comment (marehr) [inline include/seqan3/alphabet/composition/union_composition.hpp]: I want to be able to test those functions and not clutter and thus make the definition of those class variables unreadable

File: include/seqan3/alphabet/composition/union_composition.hpp (hunk 1 above)
Comment (h-2) [inline include/seqan3/alphabet/composition/union_composition.hpp]: True!

Maybe `has_alternative` (has_type) and `holds_alternative` (current type)? [the "alternative" name is not great, but it would add consistency]

What do you think about the other parts of the std::variant interface like std::get? [this would be another PR then ;-) ]

File: include/seqan3/alphabet/composition/union_composition.hpp (hunk 6 above)
Comment (h-2) [inline include/seqan3/alphabet/composition/union_composition.hpp]: I don't think you have to maintain order or forward for (static) member variables, it should "just work" (in contrast to typedefs which do need to be ordered).

Comment (marehr) [inline include/seqan3/alphabet/composition/union_composition.hpp]: If I move it down, I get 

```console
In file included from /home/marehr/develope/seqan3/test/../include/seqan3/alphabet/gap/gapped.hpp:45:0,
                 from /home/marehr/develope/seqan3/test/../include/seqan3/alphabet/gap/all.hpp:43,
                 from /home/marehr/develope/seqan3/test/../include/seqan3/alphabet/all.hpp:53,
                 from /home/marehr/develope/seqan3/test/alphabet/alphabet_test.cpp:38:
/home/marehr/develope/seqan3/test/../include/seqan3/alphabet/composition/union_composition.hpp:141:42: error: ‘alphabet_sum_size_v’ was not declared in this scope
     static constexpr size_t value_size = alphabet_sum_size_v<alphabet_types...>;
                                          ^~~~~~~~~~~~~~~~~~~

```

Comment (marehr) [inline include/seqan3/alphabet/composition/union_composition.hpp]: Hmm forwarding, doesn't work?

```console
In file included from /home/marehr/develope/seqan3/test/../include/seqan3/alphabet/gap/gapped.hpp:45:0,
                 from /home/marehr/develope/seqan3/test/../include/seqan3/alphabet/gap/all.hpp:43,
                 from /home/marehr/develope/seqan3/test/../include/seqan3/alphabet/all.hpp:53,
                 from /home/marehr/develope/seqan3/test/alphabet/alphabet_test.cpp:38:
/home/marehr/develope/seqan3/test/../include/seqan3/alphabet/composition/union_composition.hpp:123:27: error: constexpr static data member ‘alphabet_sum_size_v’ must have an initializer
     static constexpr auto alphabet_sum_size_v;
                           ^~~~~~~~~~~~~~~~~~~

```

File: include/seqan3/alphabet/composition/union_composition.hpp (hunk 1 above)
Comment (marehr) [inline include/seqan3/alphabet/composition/union_composition.hpp]: I could make `has_alternative` protected, since it is only used as concept-precondition.

We can think about adding more functions to imitate the STL behaviour. I'm open for that.

File: include/seqan3/alphabet/composition/union_composition.hpp [hunk 7]
```diff
@@ -461,39 +330,200 @@ class union_composition
...
+     * assert(table1.size() == 5);
+     * assert(table1[0] == 'A');
+     * assert(table1[1] == 'C');
+     * assert(table1[2] == 'G');
+     * assert(table1[3] == 'T');
+     * assert(table1[4] == '\0');
+     * ```
+     */
+    template <size_t max_value_size, typename alphabet_t>
+    static constexpr auto value_to_char_table_I(alphabet_t alphabet) noexcept
```
Comment (joergi-w) [inline include/seqan3/alphabet/composition/union_composition.hpp]: The capital `_I` is not compliant with our naming conventions. Should it be a Roman numeral and do we make an exception for this?

File: include/seqan3/alphabet/composition/union_composition.hpp [hunk 8]
```diff
@@ -461,39 +330,200 @@ class union_composition
...
+        constexpr auto table_size = 1 << (sizeof(char_type) * 8);
+        constexpr auto value_to_char = value_to_char_table();
+
+        using array_t = std::array<rank_t, table_size>;
+
+        array_t char_to_value{};
+        for (auto i = 0u; i < value_to_char.size(); ++i)
+        {
+            auto & old_entry = char_to_value[value_to_char[i]];
+            auto is_new_entry = value_to_char[0] != value_to_char[i] && old_entry == 0;
```
Comment (joergi-w) [inline include/seqan3/alphabet/composition/union_composition.hpp]: please use `bool` here instead of `auto`

File: include/seqan3/alphabet/composition/union_composition.hpp [hunk 9]
```diff
@@ -461,39 +330,200 @@ class union_composition
...
+        constexpr auto value_size = alphabet_sum_size_v<alphabet_types...>;
+        using rank_t = detail::min_viable_uint_t<value_size>;
+
+        constexpr auto table_size = 1 << (sizeof(char_type) * 8);
+        constexpr auto value_to_char = value_to_char_table();
+
+        using array_t = std::array<rank_t, table_size>;
+
+        array_t char_to_value{};
+        for (auto i = 0u; i < value_to_char.size(); ++i)
```
Comment (joergi-w) [inline include/seqan3/alphabet/composition/union_composition.hpp]: auto -> size_t

File: test/alphabet/composition/union_composition_detail_test.cpp [hunk 10]
```diff
@@ -41,8 +43,56 @@
...
+public:
+    using union_composition<alphabet_types...>::alphabet_sum_size_v;
+    using union_composition<alphabet_types...>::alphabet_max_size_v;
+    using union_composition<alphabet_types...>::alphabet_prefix_sum_sizes;
+    using union_composition<alphabet_types...>::value_to_char_table_I;
+    using union_composition<alphabet_types...>::value_to_char_table;
+    using union_composition<alphabet_types...>::char_to_value_table;
+};
+
+TEST(union_composition_detail_test, alphabet_sum_size_v)
```
Comment (joergi-w) [inline test/alphabet/composition/union_composition_detail_test.cpp]: I suggest to give short one-line comments for each test, where you state what the test examines.

File: test/alphabet/composition/union_composition_detail_test.cpp [hunk 11]
```diff
@@ -73,25 +123,28 @@ TEST(union_composition_detail_test, alphabet_prefix_sum_sizes)
     EXPECT_EQ(sizes4[3], 10);
 }
 
-TEST(union_composition_detail_test, union_composition_value_to_char_tableI)
+TEST(union_composition_detail_test, union_composition_value_to_char_table_I)
```
Comment (joergi-w) [inline test/alphabet/composition/union_composition_detail_test.cpp]: capital `_I` contradicts naming conventions

**Comment by h-2** [comment]
@marehr This seems almost done, can we discuss/address the remaining issues before it suffers from bit-rot?

**Comment by h-2** [comment]
@marehr can we finalise this as soon as your back?

File: include/seqan3/alphabet/composition/union_composition.hpp [hunk 12]
```diff
@@ -277,29 +113,57 @@ namespace seqan3
...
+     * ```cpp
+     * constexpr auto sum = alphabet_sum_size_v<dna4, gap, dna5>;
+     * assert(sum == 10);
+     * ```
+     */
+    template <typename ...alphabets_t>
+    //!\cond
+        requires (alphabet_concept<alphabets_t> && ...)
+    //!\endcond
+    static constexpr auto alphabet_sum_size_v =
```
[the remark this answers is not in the document]
Comment (marehr) [reply include/seqan3/alphabet/composition/union_composition.hpp]: I think this is okay, because this is a protected function, so it can be technical.

It is a sum of alphabet sizes.

I will rename it into `alphabet_sizes_sum_v`, which might be easier to read.

[the remark this answers is not in the document]
Comment (rrahn) [reply include/seqan3/alphabet/composition/union_composition.hpp]: Well, if it is detail then why not being completely expressive?
`sum_of_alphabet_sizes_v` as it is a meta-function that computes something anyway, right?

[the remark this answers is not in the document]
Comment (marehr) [reply include/seqan3/alphabet/composition/union_composition.hpp]: I think this style does not reflect the STL naming scheme. (they don't have _of_ containing in the names)

[the remark this answers is not in the document]
Comment (marehr) [reply include/seqan3/alphabet/composition/union_composition.hpp]: Okay they have (e.g. std::all_of, std::any_of, std::none_of), but they don't use them in the middle part.

I will rename all those meta functions as `sum_of_alphabet_sizes`.

[the remark this answers is not in the document]
Comment (rrahn) [reply include/seqan3/alphabet/composition/union_composition.hpp]: [See](http://en.cppreference.com/w/cpp/types/result_of)
I mean it could also be called `sum_up_alphabet_sizes_v`
I figured, that the meta-functions on type_traits are very well using preposition to clearly express the meaning. There is even a mf called `treat_as_floating_point_v`. So I don't see a reason why not expressively naming this mf.

[the remark this answers is not in the document]
Comment (marehr) [reply include/seqan3/alphabet/composition/union_composition.hpp]: renamed. Names are now: `(max|sum|partial_sum)_of_alphabet_sizes`

File: test/alphabet/composition/union_composition_detail_test.cpp (hunk 10 above)
Comment (marehr) [reply test/alphabet/composition/union_composition_detail_test.cpp]: @rrahn Is that really necessary? We don't do that right now.

**Comment by marehr** [comment]
@h-2 ready to merge, can you review it again?

**Comment by h-2** [comment]
It looks good in general, but I would really prefer if the functions for the tables be moved into the initialisers, like:

https://github.com/seqan/seqan3/blob/master/include/seqan3/alphabet/detail/convert.hpp#L62 or
https://github.com/seqan/seqan3/blob/master/include/seqan3/alphabet/nucleotide/dna4.hpp#L297

This improves:
1. expressiveness (the function is never exposed, because no one needs to use it)
2. locality (the function is defined the same place it is used and as a lambda)
3. compactness (we have less entities that we need to document, test, maintain)
4. consistency (all other conversion and value tables are implemented this way, too)

In fact I think we should add an article to the wiki on how and why to implement tables like this.

**Comment by marehr** [comment]
>This improves:
> 1. expressiveness (the function is never exposed, because no one needs to use it)
This not an issue, because they are protected.

>2. locality (the function is defined the same place it is used and as a lambda)
They are local in the class; protected

>3. compactness (we have less entities that we need to document, test, maintain)
I want to test this; I can't do this with lamda's. Making some of the function first-class functions (because the functionality may be used more than once) is much harder later. Missing doc, missing test, missing function signature makes this step much harder.

>4. consistency (all other conversion and value tables are implemented this way, too)
My design was older than you.

**Comment by h-2** [comment]
> My design was older than yours.

Please work on your attitude, this is a professional software project and not Kindergarten. 

The other design was implemented by at least four authors in this library in at least eleven different files; the current design here is used nowhere else. If you feel that this was a wrong way to implement it, we can discuss this and if there is support for changing it, we can change it everywhere, but just changing existing practice wherever someone feels like it creates a :hankey:  piece of software.

> This not an issue, because they are protected.
> They are local in the class; protected

private/protected or ::detail do not mean that everyone can do whatever they like. This is functionality exposed to the developer and it is not self explanatory, see my inline comments.

> I want to test this; I can't do this with lamda's. Making some of the function first-class functions (because the functionality may be used more than once) is much harder later. Missing doc, missing test, missing function signature makes this step much harder.

This is a valid argument, if the code is actually used somewhere else, or if you argue where it would be used, I see neither here. Even then it is highly internal code and the result should be tested, not some intermediate step.
See also "Basic Rules", Rule number 1 in our wiki, since May 2017 if dates are important:
https://github.com/seqan/seqan3/wiki/Tests "test the publicly specified interfaces"

File: include/seqan3/alphabet/composition/union_composition.hpp [hunk 13]
```diff
@@ -461,39 +330,202 @@ class union_composition
...
+     * assert(table1[0] == 'A');
+     * assert(table1[1] == 'C');
+     * assert(table1[2] == 'G');
+     * assert(table1[3] == 'T');
+     * assert(table1[4] == '\0');
+     * ```
+     */
+    template <size_t table_size, typename alphabet_t>
+        requires (alphabet_size_v<alphabet_t> <= table_size)
+    static constexpr auto make_value_to_char_table(alphabet_t alphabet) noexcept
```
Comment (h-2) [inline include/seqan3/alphabet/composition/union_composition.hpp]: This is called `make_value_to_char_table` so does it make the `value_to_char_table` ?

File: include/seqan3/alphabet/composition/union_composition.hpp [hunk 14]
```diff
@@ -461,39 +330,202 @@ class union_composition
...
+     * assert(value_to_char[1] == 'C');
+     * assert(value_to_char[2] == 'G');
+     * assert(value_to_char[3] == 'T');
+     * assert(value_to_char[4] == '-');
+     * assert(value_to_char[5] == 'A');
+     * assert(value_to_char[6] == 'C');
+     * // and so on
+     * ```
+     */
+    static constexpr auto value_to_char_table() noexcept
```
Comment (h-2) [inline include/seqan3/alphabet/composition/union_composition.hpp]: This is called `value_to_char_table`, but is actually a function, not a table :thinking: 
Also the previous function actually doesn't "make" this, it is the other way around, the previous function is used by this one.

File: include/seqan3/alphabet/composition/union_composition.hpp [hunk 15]
```diff
@@ -461,39 +330,202 @@ class union_composition
...
+        return char_to_value;
+    }
+
+    //!\brief Compile-time generated lookup table which contains the prefix sum up to the position of each alphabet.
+    //!\sa partial_sum_of_alphabet_sizes
+    static constexpr std::array partial_sum_sizes = partial_sum_of_alphabet_sizes<alphabet_types...>();
+
+    //!\brief Compile-time generated lookup table which maps the rank to char.
+    //!\sa value_to_char_table
+    static constexpr std::array value_to_char = value_to_char_table();
```
Comment (h-2) [inline include/seqan3/alphabet/composition/union_composition.hpp]: A hunder lines later this is the actual table that is relevant for everything else in this file. Strangely it has the least amount of documentation and it isn't called table, although it's the only thing where it might make sense to call it table, because it is actually data structure.

File: include/seqan3/alphabet/composition/union_composition.hpp (hunk 14 above)
Comment (h-2) [reply include/seqan3/alphabet/composition/union_composition.hpp]: Also the documentation of this actually covers `value_to_char`...

File: include/seqan3/alphabet/composition/union_composition.hpp (hunk 13 above)
Comment (marehr) [reply include/seqan3/alphabet/composition/union_composition.hpp]: It makes a `value_to_char_table` for a specific alphabet, but with a fixed-sized table. It is similar to https://github.com/seqan/seqan3/blob/b9eb19c183160f58da806cb619fd66c1519a5ee5/include/seqan3/alphabet/detail/convert.hpp#L66.

The name was `value_to_char_table_I` before, because it will be called in the i-th "iterator" of `value_to_char_table()`. Jörg thought this name was non conforming so I changed it to `make_value_to_char_table`, because it creates a fixed-size value_to_char_table for one alphabet.

File: include/seqan3/alphabet/composition/union_composition.hpp (hunk 14 above)
Comment (marehr) [reply include/seqan3/alphabet/composition/union_composition.hpp]: :tada: Professional attitude to :man_facepalming: :children_crossing:

This one is there to make the `value_to_char_table` of the `union_composition` and uses `make_value_to_char_table` in a expansion expression for every alphabet.

**Comment by marehr** [comment]
> > My design was older than yours.
>
> Please work on your attitude, this is a professional software project and not Kindergarten. 

I forgot a JONKING smiley :grimacing:  This was never meant to be a valid argument, sorry :bowing_man: 

> The other design was implemented by at least four authors in this library in at least eleven different files; the current design here is used nowhere else. If you feel that this was a wrong way to implement it, we can discuss this and if there is support for changing it, we can change it everywhere, but just changing existing practice wherever someone feels like it creates a :hankey:  piece of software.

Only that you introduced lambdas for this, before that we had manually written tables. In general they are okay.  But in this case I need to merge multiple alphabets into one. For this I use a pack expansion. To capture the pack I need a helper function (in this case `value_to_char_table`). The second function `make_value_to_char_table(_by_alphabet?)` is needed for a generating pack expansion in `value_to_char_table` to generate all the `value_to_char` tables for each alphabet. Those tables will be then merged and returned.

> > This not an issue, because they are protected.
> > They are local in the class; protected
>
> private/protected or ::detail do not mean that everyone can do whatever they like. This is functionality exposed to the developer and it is not self explanatory, see my inline comments.

Those are implementation details of the class and don't do anything excessive :question: 

> > I want to test this; I can't do this with lamda's. Making some of the function first-class functions (because the functionality may be used more than once) is much harder later. Missing doc, missing test, missing function signature makes this step much harder.
>
> This is a valid argument, if the code is actually used somewhere else, or if you argue where it would be used, I see neither here. Even then it is highly internal code and the result should be tested, not some intermediate step.
See also "Basic Rules", Rule number 1 in our wiki, since May 2017 if dates are important:
https://github.com/seqan/seqan3/wiki/Tests "test the publicly specified interfaces"

I honestly don't see why you want to forbid to test `detail` code. It made me more comfortable to write the code at the beginning and you can only manage complexity if you make every step as trivial as possible. And one final `value_to_char` table was definitely not a one trivial step in this case, but a series of transformations.

**Comment by marehr** [comment]
So I use now pure lambda function's. The code is now much more readable and not as confusing, because there are less functions with obscure names. I merged the corresponding documentations and only test the resulting protected members `char_to_value`, `value_to_char` and `partial_sum_sizes`.

@h-2 Can you have a second look and merge it if everything is alright.

**Comment by marehr** [comment]
@h-2 Okay weird it runs through in one mode, but not in the other; The same for travis O.o. I think I do stuff that I'm not allowed to. Can you have a look what may have broken the current code?

**Comment by marehr** [comment]
@h-2 FYI https://gcc.gnu.org/bugzilla/show_bug.cgi?id=84684

**Comment by marehr** [comment]
!rebuild

**Comment by h-2** [comment]
@marehr did you try the reduced example with clang? I expect it would just work...

**Comment by marehr** [comment]
I did not test it, but I think it should work.

I will try a different fold expression approach that might not have some valgrind issues. 

Am 3. März 2018 15:05:22 MEZ schrieb Hannes Hauswedell <notifications@github.com>:

**Comment by marehr** [comment]
!rebuild

**Comment by marehr** [comment]
@h-2 It seems that I found a workaround for the bug. Could you review again and then merge if everything is okay?
